File: typescript/core/src/workflow.ts
/**
 * LangGraph-inspired typed workflow graph with conditional edges.
 *
 * Phase-3 sibling of the reference engine's workflow primitive. A `Workflow<S>`
 * is a state machine: **nodes** transform a typed state value and **edges** —
 * static or **conditional** — determine the next node to execute. The runner
 * starts at the entry node, applies each node then follows its outgoing edge,
 * until it reaches the `END` sentinel (or a node with no outgoing edge), then
 * returns the final state.
 *
 * Nodes may be sync or async; the runner awaits their results. A `maxSteps` cap
 * bounds execution so an intentional or accidental cycle can't loop forever.
 *
 * Standalone module — it does not touch the agent loop. The point is the seam: a
 * multi-step orchestration (parse → guardrails → retrieve → compose → …) drops in
 * as a graph of named nodes with the routing made explicit.
 */

/** Sentinel a conditional router can return to signal termination. */
export const END = '__end__' as const;

/** A node transforms state into a new state; may be sync or async. */
export type NodeFn<S> = (state: S) => S | Promise<S>;

/** A conditional router inspects state and returns the next node name (or `END`). */
export type Router<S> = (state: S) => string;

/** Thrown when a workflow is misconfigured or exceeds its step limit. */
export class WorkflowError extends Error {
    constructor(message: string) {
        super(message);
        this.name = 'WorkflowError';
    }
}

// An edge is either a static target node name, a conditional router, or END.
type Edge<S> = { kind: 'node'; to: string } | { kind: 'conditional'; router: Router<S> } | { kind: 'end' };

/**
 * A typed workflow graph: named nodes connected by static/conditional edges.
 *
 * Build with `addNode`, `addEdge` / `addConditionalEdge`, `setEntry`, and
 * `setEnd`; the builder methods return `this` so they chain. `run` executes the
 * graph from the entry node.
 */
export class Workflow<S> {
    private readonly nodes = new Map<string, NodeFn<S>>();
    private readonly edges = new Map<string, Edge<S>>();
    private entry: string | undefined;

    constructor(private readonly maxSteps = 100) {}

    /** Register a node `func` under `name` (used to reference it in edges). */
    addNode(name: string, func: NodeFn<S>): this {
        this.nodes.set(name, func);
        return this;
    }

    /** Add a static edge `from` → `to`. */
    addEdge(from: string, to: string): this {
        this.edges.set(from, { kind: 'node', to });
        return this;
    }

    /**
     * Add a conditional edge whose `router` picks the next node at runtime. The
     * router returns the target node name, or `END` to terminate the workflow.
     */
    addConditionalEdge(from: string, router: Router<S>): this {
        this.edges.set(from, { kind: 'conditional', router });
        return this;
    }

    /** Set the entry node (first node to execute). */
    setEntry(name: string): this {
        this.entry = name;
        return this;
    }

    /** Mark `from` as terminal — reaching it ends the workflow. */
    setEnd(from: string): this {
        this.edges.set(from, { kind: 'end' });
        return this;
    }

    /**
     * Execute the workflow from the entry node, returning the final state.
     *
     * Throws {@link WorkflowError} if no entry node was set, a referenced node
     * does not exist, or the `maxSteps` cap is exceeded (e.g. an unbroken cycle).
     */
    async run(initialState: S): Promise<S> {
        if (this.entry === undefined) {
            throw new WorkflowError('workflow has no entry node — call setEntry()');
        }
        if (!this.nodes.has(this.entry)) {
            throw new WorkflowError(`entry node '${this.entry}' not found in registered nodes`);
        }

        let state = initialState;
        let current = this.entry;

        for (let step = 0; step < this.maxSteps; step++) {
            const node = this.nodes.get(current);
            if (node === undefined) {
                throw new WorkflowError(`node '${current}' not found in workflow`);
            }

            state = await node(state);

            const edge = this.edges.get(current);
            if (edge === undefined || edge.kind === 'end') {
                // No outgoing edge, or an explicit END — terminate.
                return state;
            }
            if (edge.kind === 'conditional') {
                const target = edge.router(state);
                if (target === END) return state;
                current = target;
            } else {
                current = edge.to;
            }
        }

        throw new WorkflowError(`workflow exceeded maxSteps (${this.maxSteps}) — possible infinite loop`);
    }
}

File: go/core/workflow.go
package core

import (
	"context"
	"fmt"
)

// LangGraph-inspired typed workflow graph with conditional edges.
//
// Phase-3 sibling of the reference engine's workflow primitive. A Workflow[S] is
// a state machine: nodes transform a typed state value and edges — static or
// conditional — determine the next node to execute. The runner starts at the
// entry node, applies each node then follows its outgoing edge, until it reaches
// the END sentinel (or a node with no outgoing edge), then returns the final
// state. A maxSteps cap bounds execution so an intentional or accidental cycle
// can't loop forever.
//
// Standalone — it does not touch the agent loop. The point is the seam: a
// multi-step orchestration (parse → guardrails → retrieve → compose → …) drops in
// as a graph of named nodes with the routing made explicit.

// END is the sentinel a conditional router returns to signal termination.
const END = "__end__"

// NodeFn transforms state into a new state. It receives a context so nodes can do
// async/IO work, and may return an error to abort the run.
type NodeFn[S any] func(ctx context.Context, state S) (S, error)

// Router inspects the current state and returns the next node name (or END).
type Router[S any] func(state S) string

// edgeKind discriminates the edge variants.
type edgeKind int

const (
	edgeNode edgeKind = iota
	edgeConditional
	edgeEnd
)

type edge[S any] struct {
	kind   edgeKind
	to     string
	router Router[S]
}

// Workflow is a typed workflow graph: named nodes connected by static/conditional
// edges. Build with AddNode, AddEdge / AddConditionalEdge, SetEntry, and SetEnd;
// the builder methods return the receiver so they chain. Run executes the graph.
type Workflow[S any] struct {
	nodes    map[string]NodeFn[S]
	edges    map[string]edge[S]
	entry    string
	hasEntry bool
	maxSteps int
}

// NewWorkflow creates an empty workflow with the given max-steps cap (use a value
// <= 0 for the default of 100).
func NewWorkflow[S any](maxSteps int) *Workflow[S] {
	if maxSteps <= 0 {
		maxSteps = 100
	}
	return &Workflow[S]{
		nodes:    map[string]NodeFn[S]{},
		edges:    map[string]edge[S]{},
		maxSteps: maxSteps,
	}
}

// AddNode registers a node under name (used to reference it in edges).
func (w *Workflow[S]) AddNode(name string, fn NodeFn[S]) *Workflow[S] {
	w.nodes[name] = fn
	return w
}

// AddEdge adds a static edge from → to.
func (w *Workflow[S]) AddEdge(from, to string) *Workflow[S] {
	w.edges[from] = edge[S]{kind: edgeNode, to: to}
	return w
}

// AddConditionalEdge adds a conditional edge whose router picks the next node at
// runtime. The router returns the target node name, or END to terminate.
func (w *Workflow[S]) AddConditionalEdge(from string, router Router[S]) *Workflow[S] {
	w.edges[from] = edge[S]{kind: edgeConditional, router: router}
	return w
}

// SetEntry sets the entry node (first node to execute).
func (w *Workflow[S]) SetEntry(name string) *Workflow[S] {
	w.entry = name
	w.hasEntry = true
	return w
}

// SetEnd marks from as terminal — reaching it ends the workflow.
func (w *Workflow[S]) SetEnd(from string) *Workflow[S] {
	w.edges[from] = edge[S]{kind: edgeEnd}
	return w
}

// Run executes the workflow from the entry node, returning the final state.
//
// It returns an error if no entry node was set, a referenced node does not exist,
// a node fails, or the maxSteps cap is exceeded (e.g. an unbroken cycle).
func (w *Workflow[S]) Run(ctx context.Context, initialState S) (S, error) {
	var zero S
	if !w.hasEntry {
		return zero, fmt.Errorf("workflow has no entry node — call SetEntry()")
	}
	if _, ok := w.nodes[w.entry]; !ok {
		return zero, fmt.Errorf("entry node %q not found in registered nodes", w.entry)
	}

	state := initialState
	current := w.entry

	for step := 0; step < w.maxSteps; step++ {
		node, ok := w.nodes[current]
		if !ok {
			return zero, fmt.Errorf("node %q not found in workflow", current)
		}

		next, err := node(ctx, state)
		if err != nil {
			return zero, fmt.Errorf("node %q failed: %w", current, err)
		}
		state = next

		e, ok := w.edges[current]
		if !ok || e.kind == edgeEnd {
			// No outgoing edge, or an explicit END — terminate.
			return state, nil
		}
		if e.kind == edgeConditional {
			target := e.router(state)
			if target == END {
				return state, nil
			}
			current = target
			continue
		}
		current = e.to
	}

	return zero, fmt.Errorf("workflow exceeded maxSteps (%d) — possible infinite loop", w.maxSteps)
}
